micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsCommandListener.java

```diff
@@ -56,32 +82,123 @@ public MongoMetricsCommandListener(MeterRegistry registry) {
     }
 
     @Override
-    public void commandStarted(CommandStartedEvent commandStartedEvent) {
+    public void commandStarted(CommandStartedEvent event) {
+        if (!includeCollectionNameTag) {
+            return;
+        }
+        mongoCommandUtil.determineCollectionName(event.getCommandName(), event.getCommand())
+                .ifPresent(collectionName -> addCollectionNameToCache(event, collectionName));
     }
 
     @Override
     public void commandSucceeded(CommandSucceededEvent event) {
-        timeCommand(event, event.getElapsedTime(TimeUnit.NANOSECONDS));
+        timeCommand(event, removeCollectionNameFromCache(event), event.getElapsedTime(TimeUnit.NANOSECONDS));
     }
 
     @Override
     public void commandFailed(CommandFailedEvent event) {
-        timeCommand(event, event.getElapsedTime(TimeUnit.NANOSECONDS));
+        timeCommand(event, removeCollectionNameFromCache(event), event.getElapsedTime(TimeUnit.NANOSECONDS));
     }
 
-    private void timeCommand(CommandEvent event, long elapsedTimeInNanoseconds) {
-        Timer.builder("mongodb.driver.commands")
-                .description("Timer of mongodb commands")
-                .tags(tagsProvider.commandTags(event))
-                .register(registry)
+    private void timeCommand(CommandEvent event, String collection, long elapsedTimeInNanoseconds) {
+        Timer.Builder timerBuilder = Timer.builder(this.timerMetricName)
+                .description(this.timerMetricDesc)
+                .tags(this.tagsProvider.commandTags(event));
+        if (this.includeCollectionNameTag) {
+            timerBuilder.tag("collection", collection);
+        }
+        timerBuilder
+                .register(this.registry)
                 .record(elapsedTimeInNanoseconds, TimeUnit.NANOSECONDS);
     }
 
+    private void addCollectionNameToCache(CommandEvent event, String collectionName) {
+        if (collectionNamesCache.size() < collectionNamesCacheMaxSize) {
+            collectionNamesCache.put(event.getRequestId(), collectionName);
+            return;
+        }
+        if (collectionNamesCacheOverflowErrorTracker == null) {
+            return;
+        }
+        // Cache over capacity - log every ~Nth for a good balance of warning and not spamming the logs.
+        collectionNamesCacheOverflowErrorTracker.maybeInvoke(() -> LOG.warn("Collection names cache is full - Mongo is not calling listeners properly"));
+    }
+
+    private String removeCollectionNameFromCache(CommandEvent event) {
+        return Optional.ofNullable(collectionNamesCache.remove(event.getRequestId()))
+                .orElse("unknown");
+    }
+
+    public static Builder builder(MeterRegistry registry) {
+        return new Builder(registry);
+    }
+
+    public static class Builder {
+        private final MeterRegistry registry;
+        private String timerMetricName = "mongodb.driver.commands";
```

Comment:
I would keep these hard-coded this to ensure consistency: if people start renaming these with a lot of services, it is guaranteed that they will end-up having different for the same thing. If they really want to rename them, they can do it using a `MeterFilter`.